### DISCUSSION TITLE
How to use the example demo propagator?

### DISCUSSION THREAD

**Discussion by danhngo-lx** [open-telemetry/opentelemetry-java-instrumentation#5070] [Q&A] [answered]
Hi,

I'm trying understand [the extension](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension), specifically the demo propagator.
I've built my jar file from the example (I haven't changed anything yet), then configured my service to use this extension:
```
            -javaagent:/app/opentelemetry_javaagent-1.9.1.jar
            -Dotel.javaagent.extensions=/app/opentelemetry-java-instrumentation-extension-demo-1.0-all.jar
```

Add added this env var:
```
        - name: OTEL_PROPAGATORS
          value: b3multi,demo
```

So I was expecting to see an attribute named `X-demo-field` (https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/src/main/java/com/example/javaagent/DemoPropagator.java#L24) but the there's no such attribute in the traces.
Is there anything else I need to do to configure this demo propagator?

**Comment by mateuszrzeszutek** [comment, marked as answer]
Hey @danhngo-lx ,
A custom propagator will add headers to outgoing messages (and read headers in incoming messages), but it does not add any attributes to any span by itself.
Is there something specific that you need to achieve?

**Reply by danhngo-lx** [reply]
Hi Mateusz,
Thank you for your reply.

What I'm trying achieve is to use the custom propagator to propagate a specific header from a HTTP request to all the services involved in that request, and I'm expecting to see that header in Tempo, like these attributes (tags):

![image](https://user-images.githubusercontent.com/91443857/148925696-d1b9f203-e037-4d15-94b0-d266def2f13e.png)

So is that possible with a custom propagator or I should look at something else?

**Reply by danhngo-lx** [reply]
@mateuszrzeszutek By the way, is `demo` the correct name of the propagator? It is required to specify the custom propagator name or we don't need to?

**Reply by mateuszrzeszutek** [reply]
> By the way, is `demo` the correct name of the propagator? It is required to specify the custom propagator name or we don't need to?

Yes, it's `demo` - you can see it [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/src/main/java/com/example/javaagent/DemoPropagatorProvider.java#L28).
Yes, it is required to specify the custom propagator if you want to use it.

Anyway, from what you described in your previous post I suspect that a propagator won't solve your use case -- since a propagator cannot add attributes to spans produced by the downstream services; it just propagates the context.

Instead, consider using [`Baggage`](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/baggage/Baggage.java) - you can add whatever info you need to the baggage, and it will automatically get propagated across services (assuming that you enable the `baggage` propagator, it is enabled by default by the way).
The values you put into the `Baggage` will not get automatically added to spans, but you can implement a `SpanProcessor` that does that, e.g.
```java
public void onStart(Context context, ReadWriteSpan span) {
  Baggage.fromContext(context).forEach((key, entry) ->
      span.setAttribute(key, entry.getValue());
}
```

**Reply by danhngo-lx** [reply]
Thank you very much Mateusz 😄 
I'm going to give baggage a try.
